### ISSUE TITLE
parameter identification

### ISSUE THREAD

**Issue by waynezw0618** [tum-pbs/ConFIG#12]
hi @qiauil:
sorry again. I actually works out how to use ConFIG to carry out inverse problem for Burgers problem. the results looks good, the identified nu is about 2% error better than with Adams only,  my solution is like:

```
class BurgersNN(nn.Module):
    ....

class PINN:
    def __init__(self):
        self.net = BurgersNN()
        self.nu= nn.Parameter(torch.tensor([-6.0]))

        self.optimizer_adam = torch.optim.Adam([
            {'params': self.net.parameters()},
            {'params': self.nu, 'lr': 0.01}  # 参数单独设置学习率
        ], lr=0.001)
      ....
     
     def physics_loss(self, x, t):
      ....
 
     def train(self, x_data, t_data, u_data, epochs=5000):
        for epoch in range(epochs):
            ....            
            # data loss
            self.optimizer_adam.zero_grad()
            u_pred = self.net(x_data, t_data)
            data_loss = torch.mean((u_pred - u_data)**2)
            data_loss.backward()
            grads.append(get_gradient_vector(self.net))
            losses.append(data_loss.item())

            # pde loss
            self.optimizer_adam.zero_grad()
            pde_loss = self.physics_loss(x_data, t_data)
            pde_loss.backward()
            grads.append(get_gradient_vector(self.net))
            losses.append(pde_loss.item())

            # back propogation
            apply_gradient_vector(self.net,ConFIG_update(grads))
            self.optimizer_adam.step()
            
            if epoch % 10 == 0:
                print(f"Epoch {epoch:5d}| nu: {self.nu.item():.2e}|pde loss: {pde_loss.item():.5e}|data loss: {data_loss.item():.5e}")

```
     
And now I switch to my application of 1 input & 3 outputs with 3 odes. I just modified the code based on above. the difference is of the loss, since I have to separate the  loss of each outs and odes' residual. so in PINN, I separate loss for both data and pde residual as below:

```
class PINN:
   . ...
    def data_loss(self, input, A, B , C):
        a_pred, b_pred, c_pred = self.net(input)
        loss_a = torch.mean((A - a_pred)**2)
        loss_b = torch.mean((B - b_pred)**2)
        loss_c = torch.mean((C - c_pred)**2)

        return 0.01*loss_a, loss_b, loss_c
    
    def loss_a(self, input, A, B , C):
        loss_a, _, _ = self.data_loss(input, A, B , C)
        return loss_phi
    
    def loss_b(self, t, phi, u ,v):
        _, loss_b, _ = self.data_loss(input, A, B , C)
        return loss_b
    
    def loss_c(self, t, phi, u ,v):
        _, _, loss_c = self.data_loss(input, A, B , C)
        return loss_c

    def pde_loss(self, input):
          ....
          return 1.e+10*torch.mean(fa**2),1.e+4*torch.mean(fb**2),1.e+11*torch.mean(fc**2)
   
   def loss_fa(self, input):
        loss_a, _, _ = self.pde_loss(input, A, B , C)
        return loss_fa
    
    def loss_fb(self,  input):
       ....
    
    def loss_v(self, input):
        ...
 
```

so that during the back propagation, I do collect both two respectively:

```
for epoch in range(epochs):
            grads = []
            losses = []
            
            # data loss
            for loss_fn in (self.loss_a,self.loss_b,self.loss_c):
                self.optimizer_adam.zero_grad()
                loss_d=loss_fn(input_data, a_data, b_data ,c_data)  
                loss_d.backward()
                grads.append(get_gradient_vector(self.net))
                losses.append(loss_d.item())

            # pde loss
            for loss_fn in (self.loss_fa,self.loss_fb,self.loss_fc):
                self.optimizer_adam.zero_grad()
                loss_i=loss_fn(input_data)  
                loss_i.backward()
                grads.append(get_gradient_vector(self.net))
                losses.append(loss_i.item())

            # BP
            apply_gradient_vector(self.net,ConFIG_update(grads))
            self.optimizer_adam.step()
```
different from Burgers case(where I get total mse decrease about 5-6 order ), it gets stuck with almost no decrease of mse. I also try to manually scale the loss for each, just like in pde loss. I put some scaling factor for each based on first training. It does not work neither.

can you give me some suggestion?

Wei

BTW

I also tried standard Adams, just simply add all loss in the loop. I can see very slow decrease of residual and almost very tiny change of identified parameters.

```
for epoch in range(epochs):
      self.optimizer_adam.zero_grad()
      loss_total = add all loss
      loss_total.backward()
      self.optimizer_adam.step()
```

**Comment by qiauil** [comment]
Hi, @waynezw0618 

There are lots of challenges when training PINNs. For example, the feature of the PDEs you are trying to solve is the network design. If you read our paper's appendix, you will find that there are actually many cases that can not be solved from the optimization side. 
Thus, I can't give you any useful advice regarding the information you provided, especially considering that your optimization with Adam also doesn't work here. There are just too many possible factors that can cause your problems.

Meanwhile, please only pull new issues that are related to the ConFIG code or the manuscript. The general issues for training a PINN would not be a suitable topic to be discussed here.
